speechbrain/decoders/ctc.py

```diff
@@ -113,89 +103,82 @@ def forward_step(self, g, state, candidates=None, attn=None):
             r_prev[:, 1] = torch.cumsum(
                 self.x[0, :, :, self.blank_index], 0
             ).unsqueeze(2)
-            r_prev = r_prev.view(-1, 2, self.batch_size * self.beam_size)
-            psi_prev = 0.0
+            r_prev = r_prev.view(-1, 2, n_bh)
+            psi_prev = torch.full(
+                (n_bh, self.vocab_size), 0.0, device=self.device,
+            )
         else:
-            r_prev, psi_prev = state
+            r_prev, psi_prev = states
 
         # for partial search
         if candidates is not None:
+            # The first index of each candidate.
+            cand_offset = self.batch_index * self.vocab_size
             scoring_table = torch.full(
-                (self.batch_size * self.beam_size, self.vocab_size),
+                (n_bh, self.vocab_size),
                 -1,
                 dtype=torch.long,
                 device=self.device,
             )
             # Assign indices of candidates to their positions in the table
-            col_index = torch.arange(
-                self.batch_size * self.beam_size, device=self.device
-            ).unsqueeze(1)
+            col_index = torch.arange(n_bh, device=self.device).unsqueeze(1)
             scoring_table[col_index, candidates] = torch.arange(
                 self.num_candidates, device=self.device
             )
             # Select candidates indices for scoring
             scoring_index = (
                 candidates
-                + self.cand_offset.unsqueeze(1)
-                .repeat(1, self.beam_size)
-                .view(-1, 1)
+                + cand_offset.unsqueeze(1).repeat(1, beam_size).view(-1, 1)
             ).view(-1)
             x_inflate = torch.index_select(
                 self.x.view(2, -1, self.batch_size * self.vocab_size),
                 2,
                 scoring_index,
-            ).view(2, -1, self.batch_size * self.beam_size, self.num_candidates)
+            ).view(2, -1, n_bh, self.num_candidates)
         # for full search
         else:
             scoring_table = None
```

Comment:
We really need to add some comments here, this is some serious black magic.

Reply:
:s

Reply:
welcome to beam search with batching :-)